compiler/rustc_typeck/src/check/_match.rs

```diff
@@ -468,58 +472,74 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                let mut suggest_box = !obligations.is_empty();
-                'outer: for o in obligations {
-                    for outer_ty in &[first_ty, second_ty] {
-                        match o.predicate.kind().skip_binder() {
-                            ty::PredicateKind::Trait(t) => {
-                                let pred = ty::Binder::dummy(ty::PredicateKind::Trait(
-                                    ty::TraitPredicate {
-                                        trait_ref: ty::TraitRef {
-                                            def_id: t.def_id(),
-                                            substs: self.tcx.mk_substs_trait(*outer_ty, &[]),
-                                        },
-                                        constness: t.constness,
-                                        polarity: t.polarity,
-                                    },
-                                ));
-                                let obl = Obligation::new(
-                                    o.cause.clone(),
-                                    self.param_env,
-                                    pred.to_predicate(self.tcx),
-                                );
-                                suggest_box &= self.predicate_must_hold_modulo_regions(&obl);
-                                if !suggest_box {
-                                    // We've encountered some obligation that didn't hold, so the
-                                    // return expression can't just be boxed. We don't need to
-                                    // evaluate the rest of the obligations.
-                                    break 'outer;
-                                }
+            Expectation::ExpectHasType(expected) => {
+                let TypeVariableOrigin {
+                    span,
+                    kind: TypeVariableOriginKind::OpaqueTypeInference(rpit_def_id),
+                    ..
+                } = self.type_var_origin(expected)? else { return None; };
+
+                let sig = *self
+                    .typeck_results
+                    .borrow()
+                    .liberated_fn_sigs()
+                    .get(hir::HirId::make_owner(self.body_id.owner))?;
+
+                let substs = sig.output().walk().find_map(|arg| {
+                    if let ty::GenericArgKind::Type(ty) = arg.unpack()
+                        && let ty::Opaque(def_id, substs) = *ty.kind()
+                        && def_id == rpit_def_id
+                    {
+                        Some(substs)
+                    } else {
+                        None
+                    }
+                })?;
+
+                if !self.can_coerce(first_ty, expected) || !self.can_coerce(second_ty, expected) {
+                    return None;
+                }
+
+                for ty in [first_ty, second_ty] {
+                    for pred in self.tcx.bound_explicit_item_bounds(rpit_def_id).transpose_iter() {
```

Comment:
this assumes that `bound_explicit_item_bounds` always returns `Opaque: Trait` and `<Opaque as Trait>::Assoc == Something`. While this might be the case now, I am not sure whether we implicitly use that assumption anywhere else rn.

Can you add `assert!(trait_pred.trait_ref.substs[0] == our_opaque_type)`? at least I think that's how this should work.